softlayer/vm/registry_agent_env_service_test.go

```diff
@@ -0,0 +1,179 @@
+package vm_test
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	bslcvm "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/vm"
+	boshlog "github.com/cloudfoundry/bosh-utils/logger"
+)
+
+var _ = Describe("RegistryAgentEnvService", func() {
+	var (
+		logger               boshlog.Logger
+		agentEnvService      bslcvm.AgentEnvService
+		registryServer       *registryServer
+		expectedAgentEnv     bslcvm.AgentEnv
+		expectedAgentEnvJSON []byte
+	)
+
+	BeforeEach(func() {
+		logger = boshlog.NewLogger(boshlog.LevelNone)
+
+		registryOptions := bslcvm.RegistryOptions{
+			Host:     "127.0.0.1",
+			Port:     6307,
+			Username: "fake-username",
+			Password: "fake-password",
+		}
+
+		if registryServer == nil {
+			registryServer = NewRegistryServer(registryOptions)
+
+			readyCh := make(chan struct{})
+			stopCh := make(chan error)
+
+			go func() {
+				err := registryServer.Start(readyCh)
+				if err != nil {
+					stopCh <- err
+				}
+			}()
+
+			select {
+			case <-readyCh:
+				// ok, continue
+			case err := <-stopCh:
+				panic(fmt.Sprintf("Error occurred waiting for registry server to start: %s", err))
+			}
+		}
+
+		instanceID := "fake-instance-id"
+		agentEnvService = bslcvm.NewRegistryAgentEnvService(registryOptions, instanceID, logger)
+
+		expectedAgentEnv = bslcvm.AgentEnv{AgentID: "fake-agent-id"}
+		var err error
+		expectedAgentEnvJSON, err = json.Marshal(expectedAgentEnv)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		// Leaking registry server on purpose
+		registryServer.Stop()
+	})
+
+	Describe("Fetch", func() {
+		Context("when settings for the instance exist in the registry", func() {
+			BeforeEach(func() {
+				registryServer.InstanceSettings = expectedAgentEnvJSON
+			})
+
+			It("fetches settings from the registry", func() {
+				agentEnv, err := agentEnvService.Fetch()
+				Expect(err).ToNot(HaveOccurred())
+				Expect(agentEnv).To(Equal(expectedAgentEnv))
+			})
+		})
+
+		Context("when settings for instance do not exist", func() {
+			It("returns an error", func() {
+				agentEnv, err := agentEnvService.Fetch()
+				Expect(err).To(HaveOccurred())
+				Expect(agentEnv).To(Equal(bslcvm.AgentEnv{}))
+			})
+		})
+	})
+
+	Describe("Update", func() {
+		It("updates settings in the registry", func() {
+			Expect(registryServer.InstanceSettings).To(BeNil())
+			err := agentEnvService.Update(expectedAgentEnv)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(registryServer.InstanceSettings).To(Equal(expectedAgentEnvJSON))
+		})
+	})
+})
+
+type registryServer struct {
+	InstanceSettings []byte
+	options          bslcvm.RegistryOptions
+	listener         net.Listener
+}
+
+func NewRegistryServer(options bslcvm.RegistryOptions) *registryServer {
```

Comment:
Why are you manually creating a test server instead of using the ghttp support that's already baked into ginkgo/gomega? There's a lot there that you can reuse.